### PR TITLE
String concatenation: automatic code fixes

### PR DESCRIPTION
**Pull request by akuleshov7**
Adding logic and tests for fixing string concatenation

### What's done:
1) Adding fix logic (not supported fully yet)
2) Adding tests
3) Making changes in tests to support several rules

### CODE REVIEW

File: diktat-rules/src/test/resources/test/paragraph1/naming/file/fileNameTest.kt [hunk 1]
```diff
@@ -1,12 +0,0 @@
-package org.cqfn.diktat.resources.test.paragraph1.naming.file.resources.test.paragraph1.naming.file.naming.file
```
Comment (akuleshov7) [inline diktat-rules/src/test/resources/test/paragraph1/naming/file/fileNameTest.kt:1]: very strange - looks like idea decided to remove these files, I will have a look

File: diktat-rules/src/test/resources/test/chapter1/naming/file/fileNameTest.kt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package org.cqfn.diktat.resources.test.chapter1.naming.file.resources.test.chapter1.naming.file.naming.file
```
Comment (akuleshov7) [inline diktat-rules/src/test/resources/test/chapter1/naming/file/fileNameTest.kt:1]: I do not know why, but for these files git decided that it was deleted and added not renamed

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringConcatenationRule.kt [hunk 3]
```diff
@@ -71,4 +88,63 @@ class StringConcatenationRule : Rule("string-concatenation") {
         return operationReference
                 ?.getFirstChildWithType(PLUS) != null
     }
+
+    private fun fixBinaryExpressionWithConcatenation(node: ASTNode?): ASTNode? {
+        return node?.let {
+            val binaryExpressionNode = node.psi as KtBinaryExpression
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringConcatenationRule.kt:94]:
```suggestion
            val binaryExpressionPsi = node.psi as KtBinaryExpression
```
I suggest not to use 'node' in the name of variable which isn't `ASTNode`.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringConcatenationRule.kt [hunk 4]
```diff
@@ -71,4 +88,63 @@ class StringConcatenationRule : Rule("string-concatenation") {
...
+            val parentNode = node.treeParent
+
+            // this is a special very tricky hack: we have special utility methods (KotlinParser)
+            // that let us to create a node from a string automatically by this utility
+            // instead of creating a node in AST manually by a programmer and creating an hierarchy of nodes in a tree
+
+            // FixMe: code duplication here is temporary until fix will work properly
+            if (binaryExpressionNode.isRvalueConstantExpression() || binaryExpressionNode.isRvalueStringTemplateExpression()) {
+                // =========== "a " + "b" -> "a b"
+                val lvalueText = binaryExpressionNode.left!!.text.replace("\"", "")
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringConcatenationRule.kt:105]:
```suggestion
                val lvalueText = binaryExpressionNode.left!!.text.trim("\"")
```
If there are escaped quotes in the string, you will remove them with `replace`.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringConcatenationRule.kt [hunk 5]
```diff
@@ -71,4 +88,63 @@ class StringConcatenationRule : Rule("string-concatenation") {
...
+                return newNode
+            } else if (binaryExpressionNode.isRvalueDotQualifiedExpression()) {
+                // ===========  "string " + "valueStr".replace("my", "") -> "string ${"valueStr".replace("my", "")}""
+                val lvalueText = binaryExpressionNode.left!!.text.replace("\"", "")
+                val rvalueText = binaryExpressionNode.right!!.text
+                val newNode = KotlinParser().createNode("\"$lvalueText\${$rvalueText}\"")
+                parentNode.replaceChild(node, newNode)
+                return newNode
+            } else if (binaryExpressionNode.isRvalueParenthesized()) {
+                // ===========  "string " + (1 + 2) -> "string ${(1 + 2)}"
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringConcatenationRule.kt:125]: Seems like parenthesis can be safely removed here, or do you think it would be to aggressive fixing?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringConcatenationRule.kt [hunk 6]
```diff
@@ -46,19 +49,33 @@ class StringConcatenationRule : Rule("string-concatenation") {
...
+     */
+    private fun detectStringConcatenation(topLevelBinaryExpr: ASTNode) {
+        val allBinaryExpressions = topLevelBinaryExpr.findAllDescendantsWithSpecificType(BINARY_EXPRESSION)
+        val nodeWithBug = allBinaryExpressions.find { detectStringConcatenationInExpression(it) }
+
+        val bugDetected = nodeWithBug != null
+
+        if (bugDetected) {
+            STRING_CONCATENATION.warnAndFix(configRules, emitWarn,
+                    this.isFixMode, topLevelBinaryExpr.text, nodeWithBug!!.startOffset) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/StringConcatenationRule.kt:63]:
```suggestion
                    this.isFixMode, topLevelBinaryExpr.text.lines().first(), nodeWithBug!!.startOffset) {
```
In case expression is too long and complex, I suggest to trim it for error message.

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/StringConcatenationRuleFixTest.kt [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+import org.junit.jupiter.api.Tag
+import org.junit.jupiter.api.Test
+
+class StringConcatenationRuleFixTest : FixTestBase("test/chapter3/strings", listOf(StringConcatenationRule(), DummyWarning()),
+        listOf(
+                RulesConfig(Warnings.STRING_CONCATENATION.name, true, mapOf())
+        )
+) {
+    @Test
+    @Disabled("ignored due to not all cases supported now")
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/StringConcatenationRuleFixTest.kt:19]: I think it's better to split this test into working examples and disabled unsupported examples.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt [hunk 8]
```diff
@@ -36,31 +48,166 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
...
-    private fun detectStringConcatenationInExpression(node: ASTNode, parentNode: ASTNode): Boolean {
-        assert(node.elementType == BINARY_EXPRESSION)
-        val firstChild = node.firstChildNode
-        return if (isPlusBinaryExpression(node) && firstChild.elementType == STRING_TEMPLATE) {
-            STRING_CONCATENATION.warn(configRules, emitWarn, this.isFixMode, parentNode.text, firstChild.startOffset, firstChild)
-            true
-        } else {
-            false
+    private fun circle(parentTopLevelBinaryExpr: ASTNode) {
+        val allBinaryExpressions = parentTopLevelBinaryExpr.findAllDescendantsWithSpecificType(BINARY_EXPRESSION)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt:71]: You are performing the same logic twice for top-level expression - while checking if warning should be raised and at first iteration. This can be refactored

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt [hunk 9]
```diff
@@ -36,31 +48,166 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
...
-    @Suppress("FUNCTION_BOOLEAN_PREFIX")
-    private fun detectStringConcatenationInExpression(node: ASTNode, parentNode: ASTNode): Boolean {
-        assert(node.elementType == BINARY_EXPRESSION)
-        val firstChild = node.firstChildNode
-        return if (isPlusBinaryExpression(node) && firstChild.elementType == STRING_TEMPLATE) {
-            STRING_CONCATENATION.warn(configRules, emitWarn, this.isFixMode, parentNode.text, firstChild.startOffset, firstChild)
-            true
-        } else {
-            false
+    private fun circle(parentTopLevelBinaryExpr: ASTNode) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt:70]: I think, loop is better name

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt [hunk 10]
```diff
@@ -36,31 +48,166 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
...
         }
     }
 
-    @Suppress("COMMENT_WHITE_SPACE")
+    /**
+     * We can detect string concatenation by the first (left) operand in binary expression.
+     * If it is of type string - then we found string concatenation.
+     */
+    private fun isDetectStringConcatenationInExpression(node: ASTNode): Boolean {
+        assert(node.elementType == BINARY_EXPRESSION) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt:86]:
```suggestion
        require(node.elementType == BINARY_EXPRESSION) {
```
`assert` is JVM-only and has some restrictions, while `require` does the same, but is native to Kotlin :)

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt [hunk 11]
```diff
@@ -36,31 +48,166 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
...
+        val firstChild = node.firstChildNode
+        return (isPlusBinaryExpression(node) &&
+                (firstChild.elementType == STRING_TEMPLATE ||
+                        (firstChild.text.contains("toString()")) && firstChild.elementType == DOT_QUALIFIED_EXPRESSION)
+        )
+    }
+
     private fun isPlusBinaryExpression(node: ASTNode): Boolean {
         assert(node.elementType == BINARY_EXPRESSION)
-        //     binary expression
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt:58]: Please suppress `COMMENT_WHITE_SPACE` and keep this formatting

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt [hunk 12]
```diff
@@ -36,31 +48,166 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
...
+     * If it is of type string - then we found string concatenation.
+     */
+    private fun isDetectStringConcatenationInExpression(node: ASTNode): Boolean {
+        assert(node.elementType == BINARY_EXPRESSION) {
+            "cannot process non binary expression in the process of detecting string concatenation"
+        }
+        val firstChild = node.firstChildNode
+        return (isPlusBinaryExpression(node) &&
+                (firstChild.elementType == STRING_TEMPLATE ||
+                        (firstChild.text.contains("toString()")) && firstChild.elementType == DOT_QUALIFIED_EXPRESSION)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt:92]:
```suggestion
                        (firstChild.text.endsWith("toString()")) && firstChild.elementType == DOT_QUALIFIED_EXPRESSION)
```
There can be cases when resulting string is transformed into something else

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt [hunk 13]
```diff
@@ -36,31 +48,166 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
...
+        // /        |        \
+        // expr1 operationRef expr2
 
         val operationReference = node.getFirstChildWithType(OPERATION_REFERENCE)
         return operationReference
             ?.getFirstChildWithType(PLUS) != null
     }
+
+    private fun fixBinaryExpressionWithConcatenation(node: ASTNode?) {
+        val binaryExpressionNode = node?.psi as KtBinaryExpression
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt:108]:
```suggestion
        val binaryExpressionPsi = node?.psi as KtBinaryExpression
```
It's confusing when variable holds PSI but is named as if it holds ASTNode

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt [hunk 14]
```diff
@@ -36,31 +48,166 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
...
+
+    private fun isPlusBinaryExpressionAndFirstElementString(binaryExpressionNode: KtBinaryExpression) =
+            (binaryExpressionNode.left is KtStringTemplateExpression) && PLUS == binaryExpressionNode.operationToken
+
+    @Suppress(
+        "TOO_LONG_FUNCTION",
+        "NESTED_BLOCK",
+        "SAY_NO_TO_VAR")
+    private fun checkKtExpression(binaryExpressionNode: KtBinaryExpression): String {
+        var lvalueText = binaryExpressionNode.left!!.text.replace("\"", "")
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt:123]:
```suggestion
        var lvalueText = binaryExpressionNode.left!!.text.trim("\"")
```
In case there are escaped quoates inside the string, or just quotes inside triple-quoted string

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt [hunk 15]
```diff
@@ -36,31 +48,166 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
...
+        val rvalueText = binaryExpressionNode.right!!.text
+
+        if (binaryExpressionNode.isLvalueDotQualifiedExpression() && binaryExpressionNode.firstChild.text.contains("toString()")) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt:126]:
```suggestion
        if (binaryExpressionNode.isLvalueDotQualifiedExpression() && binaryExpressionNode.firstChild.text.endsWith("toString()")) {
```

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/AstNodeUtils.kt [hunk 16]
```diff
@@ -415,9 +407,23 @@ fun ASTNode.numNewLines() = text.count { it == '\n' }
...
  */
-fun ASTNode.findAllNodesWithSpecificType(elementType: IElementType, withSelf: Boolean = true) =
+fun ASTNode.findAllDescendantsWithSpecificType(elementType: IElementType, withSelf: Boolean = true) =
         findAllNodesWithCondition({ it.elementType == elementType }, withSelf)
 
+/**
+ * This method performs tree traversal and returns all nodes that match condition (lambda).
+ *
+ */
+fun ASTNode.getAllDescendantsWithLevel(level: Int = 0, condition: (ASTNode) -> Boolean): Map<ASTNode, Int> {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/AstNodeUtils.kt:417]: Why is this needed?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/StringCaseUtils.kt [hunk 17]
```diff
@@ -46,6 +46,7 @@ fun String.isLowerSnakeCase(): Boolean = this.matches("(([a-z]+)_*)+[a-z0-9]*".t
  *
  * @return converted string
  */
+@Suppress("STRING_CONCATENATION")
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/StringCaseUtils.kt:49]: Well, fix the code according to the new rule :)

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt [hunk 18]
```diff
@@ -36,31 +48,166 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
...
+            // =========== "a " + "b" -> "a b"
+            val rvalueTextNew = rvalueText.replace("\"", "")
+            return "$lvalueText$rvalueTextNew"
+        } else if (binaryExpressionNode.isRvalueReferenceExpression()) {
+            // ===========  "a " + b -> "a $b"
+            return "$lvalueText$$rvalueText"
+        } else if (binaryExpressionNode.isRvalueDotQualifiedExpression()) {
+            // ===========  "string " + "valueStr".replace("my", "") -> "string ${"valueStr".replace("my", "")}""
+            return "$lvalueText\${$rvalueText}"
+        } else if (binaryExpressionNode.isRvalueOperation()) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt:156]: This branch can be merged with the previous

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt [hunk 19]
```diff
@@ -36,31 +48,166 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
...
+            // ===========  "sum " + (1 + 2 + 3) * 4 -> "sum ${(1 + 2 + 3) * 4}"
+            return "$lvalueText\${$rvalueText}"
+        } else if (binaryExpressionNode.isRvalueParenthesized()) {
+            val binExpression = binaryExpressionNode.right!!.children.get(0)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt:160]:
```suggestion
            val binExpression = binaryExpressionNode.right!!.children.first()
```

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/StringCaseUtils.kt (hunk 17 above)
Comment (Cheshiriks) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/StringCaseUtils.kt:49]: There's a method chain that returns a string. I would not want to fix it. It will be ugly.
return this[idx].toUpperCase().toString() -> return "${this[idx].toUpperCase().toString()}"

Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/StringCaseUtils.kt:49]: But there is the expression `this[idx].toUpperCase().toString() + convertUnknownCaseToUpperSnake(this.substring(idx + 1))`, which would be converted to `"${this[idx].toUpperCase()}${convertUnknownCaseToUpperSnake(this.substring(idx + 1))}"`. My concern is, that if we start suppressing the rule right away, there may be something in the rule that needs tuning. @akuleshov7 could you suggest whether this is desired behaviour?

File: diktat-rules/src/test/resources/test/paragraph3/string_concatenation/StringConcatenationTest.kt [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+val myTest5 = "trying to sum with " + """multiline"""
+val myTest6 = "trying to sum with " + """
+    multiline""".trimIndent()
+val myTest7 = """multiline""" + " string"
+val myTest8 = "string " + valueStr.replace("my", "")
+val myTest9 = "string " + "valueStr".replace("my", "")
+val myTest10 = "string " + (1 + 5)
+val myTest11 = "sum " + ("other string " + 3 + " str2 " + 5 + (" other string 2 " + 2 + " str3 " + 4))
+val myTest12 = "my string " + 1 + 2 + 3
+val myTest13 = (1 + 2).toString() + " my string " + 3 + " string " + valueStr + valueStr
```
Comment (akuleshov7) [inline diktat-rules/src/test/resources/test/paragraph3/string_concatenation/StringConcatenationTest.kt:18]: need to add the following test:
`a.toString() + "fghgfhfg"` should not be fixed to "$afghgfhfg", but should be fixed to "${a}fghgfhfg"

`a.toString() + " fghgfhfg"` should be fixed to "$a fghgfhfg"

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/StringCaseUtils.kt (hunk 17 above)
Comment (akuleshov7) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/StringCaseUtils.kt:49]: Let's change the Rule: if the `r-value` is not a constant string then DO NOT change them to template.
BUT if the the r-value is a constant string, it should be concatenated: `a.toString()` + " kjfgjfhgkjhdfg" -> "$a kjfgjfhgkjhdfg".

Please also reflect it in the codestyle!

File: diktat-rules/src/test/resources/test/paragraph3/string_concatenation/StringConcatenationTest.kt (hunk 20 above)
Comment (petertrr) [reply diktat-rules/src/test/resources/test/paragraph3/string_concatenation/StringConcatenationTest.kt:23]: We can make it even easier: in string concatenation rule always put braces, and then SRING_TEMPLATE_CURLY_BRACES inspection will remove redundant ones

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt [hunk 21]
```diff
@@ -36,25 +48,56 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
...
+     * If it is of type string - then we found string concatenation.
+     * If the right value is not a constant string then don't change them to template.
+     */
+    private fun isDetectStringConcatenationInExpression(node: ASTNode): Boolean {
+        require(node.elementType == BINARY_EXPRESSION) {
+            "cannot process non binary expression in the process of detecting string concatenation"
         }
+        val firstChild = node.firstChildNode
+        val lastChild = node.lastChildNode
+        return (isPlusBinaryExpression(node) && (lastChild.elementType != CALL_EXPRESSION) &&
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt:92]: What is this condition? Why `+` and  rvalue isn't a call expression?

File: diktat-rules/src/test/resources/test/paragraph3/string_concatenation/StringConcatenationTest.kt [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+val myTest14 = "my string " + (1 + 2 + 3) + (" other string " + 3) + (1 + 2 + 3)
+val myTest15 = 1 + 2 + ("1" + 3).toInt()
+val myTest16 = 1.0 + 2.0 + ("1" + 3.0).toFloat()
+val myTest17 = "sum " + (1 + 2 + 3) * 4
+val myTest18 = "my string " + (1 + 2 + 3) * 4 + (" other string " + 3) + (1 + (2 + 3)) + (" third string " + ("str " + 5))
+val myTest19 = 1 + 2 + 3 + ("6" + 5).toInt()
+val myTest20 = x.toString() + "string"
+val myTest21 = x.toString() + " string"
+val myTest22 = "string" + foo()
```
Comment (petertrr) [inline diktat-rules/src/test/resources/test/paragraph3/string_concatenation/StringConcatenationTest.kt:32]: Did we decide not to fix this? I don't remembder, why? I believe it was about `foo.toString() + bar`, i.e. when neither part is string literal

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt [hunk 23]
```diff
@@ -36,25 +49,56 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
...
+     */
+    private fun isDetectStringConcatenationInExpression(node: ASTNode): Boolean {
+        require(node.elementType == BINARY_EXPRESSION) {
+            "cannot process non binary expression in the process of detecting string concatenation"
         }
+        val firstChild = node.firstChildNode
+        val lastChild = node.lastChildNode
+        return (isPlusBinaryExpression(node) && !(firstChild.elementType != STRING_TEMPLATE && lastChild.elementType == CALL_EXPRESSION) &&
+                (firstChild.elementType == STRING_TEMPLATE ||
+                        ((firstChild.text.endsWith("toString()")) && firstChild.elementType == DOT_QUALIFIED_EXPRESSION))
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt:95]:
```suggestion
        return (isPlusBinaryExpression(node) && (firstChild.elementType == STRING_TEMPLATE ||
                        (firstChild.text.endsWith("toString()") && firstChild.elementType == DOT_QUALIFIED_EXPRESSION && lastChild.elementType != CALL_EXPRESSION))
```
I'd merge these two conditions for readability. Also, it might be worth extracting them to a variable.

**Review by petertrr — APPROVED** [review]
